### DISCUSSION TITLE
Form fields can never be copied from a source PDF when using doc.insert_pdf()

### DISCUSSION THREAD

**Discussion by ag-gaphp** [pymupdf/PyMuPDF#3315] [Q&A] [answered]
![image](https://github.com/pymupdf/PyMuPDF/assets/164041490/d4289ec9-fa30-419d-88f5-50105328928c)
On this page: https://pymupdf.readthedocs.io/en/latest/document.html#Document.insert_pdf

I'm just curious, is this a limitation of PDFs in general or specific to MuPDF?

I can accomplish what I want to accomplish by reading all the field data from one file, and then adding brand new fields to the PDF that I need them added to after inserting the blank page. Obviously it would be a tiny bit less cumbersome if form fields could just be copied directly from the source PDF.

Not expecting anything to change in PyMuPDF, just curious about why the docs mention, "Form fields can never be copied", and where that limitation comes from.

**Comment by JorjMcKie** [comment, marked as answer]
Form fields, like annotations, are specific to the page for which they are defined.
But in contrast to them, widgets also are listed document-wide in a central dictionary called `/AcroForm` in the PDF's catalog. The fact that the array `/AcroForm/Fields` is not empty determines whether the PDF is a **_Form PDF_**. Other information may also be present in AcroForm like inter-field relationships and more.
Bottom line, like all document-wide, non-page related information and objects, widget-related information cannot be copied by a page-wise operation line `.insert_pdf()` - it is impossible to convert the receiving PDF into a Form PDF.
The same is true for a number of other document-wide objects and properties (embedded files, table of contents, page labels, optional content structure elements).

**Reply by ag-gaphp** [reply]
That's interesting. There's so much more in the PDF world that I never knew about until I started my latest contract, and it's no wonder some other modules I tried before yours were really tanking my files when trying to combine pages. I guess there is some form of magic that Adobe Acrobat employs that allows me to merge PDFs together, and keep all the form data, regardless if the base doc was a form or not. They probably do a lot of leg work to make that happen, but makes sense since they wrote the book on it, literally.

If I wanted to dive in even more, is the [Adobe reference](https://opensource.adobe.com/dc-acrobat-sdk-docs/standards/pdfstandards/pdf/PDF32000_2008.pdf) the best place to start in your opinion? I'm mostly OK with technical documentation, but if you knew of any other well-organized resources that were a bit more high-level I'd love to take a look at them and learn more.

Your module is one of the most consistently functional PDF modules I've tested and started using, not to mention you actually reply to issues (even when the user is clearly doing something stupid) and have solid documentation. I appreciate your effort and work!

**Reply by JorjMcKie** [reply]
You must be referring to the paid version of Adobe Acrobat.
I didn't say that merging PDF's with keeping widgets is impossible to do, did I.
It's just something requiring a lot of work to do. So you cannot expect it to be freeware.

**Reply by JorjMcKie** [reply]
Otherwise: yes I used the reference book to get acquainted with PDF internals.
